x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/TimeSeriesSortedSourceOperatorFactory.java

```diff
@@ -228,62 +228,43 @@ protected boolean lessThan(Leaf a, Leaf b) {
             void consume() throws IOException {
                 if (queue != null) {
                     currentTsid = BytesRef.deepCopyOf(queue.top().timeSeriesHash);
-                    boolean breakOnNextTsidChange = false;
                     while (queue.size() > 0) {
-                        if (remainingDocs <= 0) {
+                        if (remainingDocs <= 0 || currentPagePos >= maxPageSize) {
                             break;
                         }
-                        if (currentPagePos > maxPageSize) {
-                            breakOnNextTsidChange = true;
-                        }
-
                         currentPagePos++;
                         remainingDocs--;
                         Leaf leaf = queue.top();
                         segmentsBuilder.appendInt(leaf.segmentOrd);
                         docsBuilder.appendInt(leaf.iterator.docID());
                         timestampIntervalBuilder.appendLong(leaf.timestamp);
                         tsOrdBuilder.appendInt(globalTsidOrd);
+                        final Leaf newTop;
                         if (leaf.nextDoc()) {
                             // TODO: updating the top is one of the most expensive parts of this operation.
                             // Ideally we would do this a less as possible. Maybe the top can be updated every N docs?
-                            Leaf newTop = queue.updateTop();
```

Comment:
The idea behind this logic is that all docs of a time serie are always contained within a page. Downstream operators could optimize for that, currently this doesn't happen and the grouping operator in the other pr doesn't either (yet). So I'm okay with simplifying this for now and at the same time fixing the bug :)

Reply:
Yes, I understand that idea and we will re-enable this short-cut by using ConstantBlocks.